pyaccel/lifetime.py

```diff
@@ -313,45 +449,179 @@ def touschek_data(self):
+
+            # Lower integration limit
+            ksip = (2*_np.sqrt(c_var)/gamma * d_accp)**2
+            ksin = (2*_np.sqrt(c_var)/gamma * d_accn)**2
+
+            # Interpolate d_touschek
+            d_pos = _np.interp(
+                ksip, self._KSI_TABLE, self._D_TABLE, left=0.0, right=0.0)
+            d_neg = _np.interp(
+                ksin, self._KSI_TABLE, self._D_TABLE, left=0.0, right=0.0)
+
+            touschek_coeffs['a_var'] = a_var
+            touschek_coeffs['b_var'] = b_var
+            touschek_coeffs['c_var'] = c_var
+            touschek_coeffs['ksip'] = ksip
+            touschek_coeffs['ksin'] = ksin
+            touschek_coeffs['d_pos'] = d_pos
+            touschek_coeffs['d_neg'] = d_neg
+
+            # Touschek rate
+            ratep = const * nr_part/gamma**2 / d_accp**3 * d_pos / vol
+            raten = const * nr_part/gamma**2 / d_accn**3 * d_neg / vol
+            rate = (ratep+raten)/2
+        elif self.touschek_model == self.TOUSCHEKMODEL.Piwinski:
+            eta1til2 = (alpha1*eta1 + beta1*eta1l)**2
+            eta2til2 = (alpha2*eta2 + beta2*eta2l)**2
+            espread2 = espread*espread
+            betagamma2 = (beta*gamma)**2
+
+            val1 = 1/espread2
+            val2 = (eta1*eta1 + eta1til2)/(sig1b2)
+            val3 = (eta2*eta2 + eta2til2)/(sig2b2)
+            sigh2 = 1/(val1 + val2 + val3)
+
+            c1_ = beta1**2/sig1b2*(1-sigh2*eta1til2/sig1b2)
+            c2_ = beta2**2/sig2b2*(1-sigh2*eta2til2/sig2b2)
+            ch_ = (sig1*sig2)**2 - (espread2*eta1*eta2)**2
+
+            b1_ = (c1_ + c2_)/(2*betagamma2)
+            b2_ = (c1_ - c2_)**2/4
+            b2_ += eta1til2*eta2til2*(sigh2/emit1/emit2)**2
+            b2_ /= (betagamma2**2)
+            b2_ = _np.sqrt(b2_)
+
+            taum_p = (beta*d_accp)**2
+            taum_n = (beta*d_accn)**2
+
+            f_int_p = self.f_integral_2_simps(taum_p, b1_, b2_)
+            f_int_n = self.f_integral_2_simps(taum_n, b1_, b2_)
+
+            touschek_coeffs['b1'] = b1_
+            touschek_coeffs['b2'] = b2_
+            touschek_coeffs['taum_p'] = taum_p
+            touschek_coeffs['taum_n'] = taum_n
+            touschek_coeffs['f_int_p'] = f_int_p
+            touschek_coeffs['f_int_n'] = f_int_n
+
+            rate_const = const * nr_part/gamma**2/bunlen
+            rate_const /= _np.sqrt(ch_)
+            ratep = rate_const * f_int_p/taum_p
+            raten = rate_const * f_int_n/taum_n
+            rate = (ratep + raten)/2
+
+        rate = _np.array(rate).ravel()
+        # Average inverse Touschek Lifetime
         avg_rate = _np.trapz(rate, x=s_calc) / (s_calc[-1] - s_calc[0])
-        return dict(rate=rate, avg_rate=avg_rate, volume=vol, pos=s_calc)
+        dit = dict(
+            rate=rate, avg_rate=avg_rate,
+            volume=vol, pos=s_calc,
+            touschek_coeffs=touschek_coeffs)
+        return dit
+
+    @staticmethod
+    def f_function_arg_1(tau, taum, b1_, b2_):
```

Comment:
I think it is a good idea to explain what this `f_function_arg_1` is. Maybe put a reference to the equation of the Piwinsky paper here...

Reply:
Ok, I deleted the implementation of the expression without the substitution (I just kept the expression with integration from kappam to pi/2, equation without number in Piwiski paper, just below Eq. 42). Therefore I used the names `f_function_arg` and `f_integral_simps` without the need to use the number 1 and 2 to differentiate the two expressions. Then, I added doctrings with references to the expressions.